# Business Problem

The Tanzania Development Trust is a UK charitable organization operating within the country of Tanzania since 1975.

They focus on development in rural Tanzania, aiming to support small projects in the poorest parts of the country where one of their priority areas of funding is clean water. Their stated water project involves boreholing and rope pump installation in areas with limited access to clean water, currently located in the regions of Kagera and Kigoma in the northwest of the country.

A new benefactor wants to expand the project not only geographically to more of the country, but in the scope of repairing existing pumps before they fail. I have been tasked with developing a model to predict the operating condition of a current waterpoint: functional, needs repair, or non-functional.

The main objective is to identify waterpoints that are in need of repair. [Research shows](https://sswm.info/entrepreneurship-resource/developing-impactful-businesses/maintenance-services-for-rural-water-pumps) that it is much less expensive to repair and rehabilitate a waterpoint, as well as being more protective of the water resources in the country. 
The secondary objective is to identify concentrations of non-functioning water points that may be an eligible location for a new installation.
______________________
The data provided for modeling was collected between March 2011 and March 2013, and contains the information for 59,400 water points 

# Imports

In [1]:
import pandas as pd

# import training data and target
raw_data = pd.read_csv('data/training_data.csv')
raw_target = pd.read_csv('data/training_target.csv')

In [2]:
display(raw_target.info())
print(raw_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            59400 non-null  int64 
 1   status_group  59400 non-null  object
dtypes: int64(1), object(1)
memory usage: 928.2+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 40 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   amount_tsh             59400 non-null  float64
 2   date_recorded          59400 non-null  object 
 3   funder                 55765 non-null  object 
 4   gps_height             59400 non-null  int64  
 5   installer              55745 non-null  object 
 6   longitude              59400 non-null  float64
 7   latitude               59400 non-null  float64
 8   wpt_name               59400 non-null  object 
 9   num_private            59400 non-null  int64  
 10  basin                  59400 non-null  object 
 11  subvillage             59029 non-null  object 
 12  region                 59400 non-null  object 
 13  region_code            59400 non-null  int64  
 14  district_code          59400 non-null  int64  
 15  lg

Create a raw dataframe with merged data and target. We will use this df during initial EDA so we can compare feature relationships with target, and so we can understand and deal with null values

In [3]:
raw_df = pd.merge(raw_data, raw_target, on='id')

The dataset for training includes 59,400 entries with 40 total features and a target label.

In [11]:
status_values = pd.DataFrame(raw_df.status_group.value_counts())
status_values['percent'] = raw_df.status_group.value_counts(normalize=True) * 100
status_values

status_group    percent
functional                      32259  54.308081
non functional                  22824  38.424242
functional needs repair          4317   7.267677

This is a ternary classification problem. The three possible values are:
- functional (F)
- non functional (NF)
- functional needs repair (FR)

Value counts show that our dataset is not balanced with respect to the label values. Only 7.3% of pumps are classified as functional needs repair, while 54.3% are functional and 38.4% are non functional. We will need to keep this imbalance in mind when modeling.


In [12]:
raw_df.region.value_counts().sort_index()

Arusha           3350
Dar es Salaam     805
Dodoma           2201
Iringa           5294
Kagera           3316
Kigoma           2816
Kilimanjaro      4379
Lindi            1546
Manyara          1583
Mara             1969
Mbeya            4639
Morogoro         4006
Mtwara           1730
Mwanza           3102
Pwani            2635
Rukwa            1808
Ruvuma           2640
Shinyanga        4982
Singida          2093
Tabora           1959
Tanga            2547
Name: region, dtype: int64

In [13]:
raw_df.extraction_type.value_counts()

gravity                      26780
nira/tanira                   8154
other                         6430
submersible                   4764
swn 80                        3670
mono                          2865
india mark ii                 2400
afridev                       1770
ksb                           1415
other - rope pump              451
other - swn 81                 229
windmill                       117
india mark iii                  98
cemo                            90
other - play pump               85
walimi                          48
climax                          32
other - mkulima/shinyanga        2
Name: extraction_type, dtype: int64

In [14]:
raw_df[raw_df.status_group == 'non functional']

id  amount_tsh date_recorded                  funder  gps_height  \
3      67743         0.0    2013-01-28                  Unicef         263   
6      19816         0.0    2012-10-01                    Dwsp           0   
7      54551         0.0    2012-10-09                   Rwssp           0   
8      53934         0.0    2012-11-03                Wateraid           0   
16     48451       500.0    2011-07-04                  Unicef        1703   
...      ...         ...           ...                     ...         ...   
59382  47002         6.0    2013-08-03               Ces(gmbh)        1383   
59384  72148         0.0    2011-07-10                 Concern           0   
59391  44885         0.0    2013-08-03  Government Of Tanzania         540   
59392  40607         0.0    2011-04-15  Government Of Tanzania           0   
59394  11164       500.0    2011-03-09              World Bank         351   

        installer  longitude   latitude              wpt_name  num_private  \
3          UNICEF  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
6            DWSP  33.362410  -3.766365            Kwa Ngomho            0   
7             DWE  32.620617  -4.226198          Tushirikiane            0   
8       Water Aid  32.711100  -5.146712     Kwa Ramadhan Musa            0   
16            DWE  34.642439  -9.106185       Kwa John Mtenzi            0   
...           ...        ...        ...                   ...          ...   
59382         DWE  37.454759  -3.323599        Kwa Luka Msaki            0   
59384     CONCERN  30.667805  -2.483710              Chamkube            0   
59391  Government  38.044070  -4.272218                   Kwa            0   
59392  Government  33.009440  -8.520888        Benard Charles            0   
59394    ML appro  37.634053  -6.124830            Chimeredya            0   

       ... water_quality quality_group      quantity  quantity_group  \
3      ...          soft          good           dry             dry   
6      ...          soft          good        enough          enough   
7      ...         milky         milky        enough          enough   
8      ...         salty         salty      seasonal        seasonal   
16     ...          soft          good           dry             dry   
...    ...           ...           ...           ...             ...   
59382  ...          soft          good        enough          enough   
59384  ...          soft          good  insufficient    insufficient   
59391  ...          soft          good        enough          enough   
59392  ...          soft          good        enough          enough   
59394  ...          soft          good        enough          enough   

             source   source_type source_class              waterpoint_type  \
3       machine dbh      borehole  groundwater  communal standpipe multiple   
6       machine dbh      borehole  groundwater                    hand pump   
7      shallow well  shallow well  groundwater                    hand pump   
8       machine dbh      borehole  groundwater                    hand pump   
16            river    river/lake      surface           communal standpipe   
...             ...           ...          ...                          ...   
59382        spring        spring  groundwater           communal standpipe   
59384        spring        spring  groundwater           communal standpipe   
59391         river    river/lake      surface           communal standpipe   
59392        spring        spring  groundwater           communal standpipe   
59394   machine dbh      borehole  groundwater           communal standpipe   

      waterpoint_type_group    status_group  
3        communal standpipe  non functional  
6                 hand pump  non functional  
7                 hand pump  non functional  
8                 hand pump  non functional  
16       communal standpipe  non functional  
...                     ...             ...  


# Null Checks

In [15]:
null_checks = pd.DataFrame(data=raw_df.isna().sum(),
                          columns=['null_count'])
null_checks['percent_of_data'] = (null_checks.null_count / len(raw_data)) * 100
null_checks = null_checks[null_checks.percent_of_data > 0.0]
null_checks.sort_values('percent_of_data', ascending=False, inplace=True)
null_checks

null_count  percent_of_data
scheme_name             28166        47.417508
scheme_management        3877         6.526936
installer                3655         6.153199
funder                   3635         6.119529
public_meeting           3334         5.612795
permit                   3056         5.144781
subvillage                371         0.624579

There are 7 features with null values in our dataset, and we can what that number of nulls is by percent of total available data. 
______
We will work our way up from the smallest null_count to the largest and decide how to deal with the missing data

## Subvillage

In [16]:
subvillage_nans = raw_df[raw_df.subvillage.isnull()]
subvillage_nans.status_group.value_counts(normalize=True)

functional                 0.552561
non functional             0.444744
functional needs repair    0.002695
Name: status_group, dtype: float64

The null values in subvillage represent 0.6% of our total data. The distribution of the target label is very close to the whole dataset

In [17]:
subvillage_nans.region.value_counts()

Dodoma    361
Mwanza     10
Name: region, dtype: int64

All but 10 of our subvillage nan's come from the region of Dodoma, the rest from Mwanza. Lets look at the subvillage distribution of those regions from the whole dataset

In [18]:
raw_df[raw_df['region'] == 'Dodoma'].subvillage.value_counts()

Kawawa         54
Shuleni        43
Nyerere        35
Azimio         34
Majengo        32
               ..
Foye            1
Mtatangwe       1
Makao Mapya     1
Soya Mjini      1
Mgomwa          1
Name: subvillage, Length: 705, dtype: int64

In [19]:
raw_df[raw_df['region'] == 'Mwanza'].subvillage.value_counts()

1                     132
Madukani               52
Bujingwa               25
Shuleni                19
Matale                 18
                     ... 
Bukalo                  1
Nyambona                1
Kabaganda B             1
Bulyahilu Center B      1
Mwambogwa               1
Name: subvillage, Length: 1507, dtype: int64